reinforcement_learning/rl_cartpole_batch_coach/rl_cartpole_batch_coach.ipynb

```diff
@@ -0,0 +1,712 @@
+{
```

Comment:
<p>Edited text below:</p><p> </p><p>For many real-world problems, the reinforcement learning (RL) agent needs to learn from historical data that was generated by some deployed policy. For example, we may have historical data of experts playing games, users interacting with a website or sensor data from a control system. This notebook shows an example of how to use batch RL to train a new policy from offline dataset. We use gym&nbsp;<code>CartPole-v0</code>&nbsp;as a fake simulated system to generate offline dataset and the RL agents are trained using Amazon SageMaker RL.</p><p> </p><p>We may want to evaluate the policy learned from historical data before deployment. Since simulators may not be available in all use cases, we need to evaluate how good the learned policy is using held out historical data. This is called as off-policy evaluation or counterfactual evaluation. We evaluate the policy using several off-policy evaluation algorithms in this notebook.  </p><p> </p><p>We can deploy the policy using SageMaker Hosting endpoint. However, some use cases may have low throughput (e.g., one interaction per hour) and low latency (e.g. 1ms) requirements, and an endpoint solution may not satisfy the requirements. Here we demonstrate how to deploy the policy with SageMaker Batch Transform, where the inferences of the policy are computed beforehand and stored offline in a lookup table for serving. Batch Transform processes input to the policy in batches and is ideal for scenarios where the feedback (in the form of next state and reward) from the environment is expected to be delayed by hours. </p><p> </p><p>Figure below shows an overview of the entire notebook &lt;add figure&gt;. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/awslabs/amazon-sagemaker-examples/pull/926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Reply:
<blockquote>We may want to evaluate the policy learned from historical data before deployment. Since simulators may not be available in all use cases, we need to evaluate how good the learned policy is using held out historical data. This is called as off-policy evaluation or counterfactual evaluation. We evaluate the policy using several off-policy evaluation algorithms in this notebook.</blockquote><p> </p><p>This might create confusion to the audience. The OPE is included as part of the training job in the notebook, while the evaluation job running in the notebook is actually interacting with the simulator, generating rewards but not OPE metrics</p>

Reply:
<blockquote>We can deploy the policy using SageMaker Hosting endpoint. However, some use cases may have low throughput (e.g., one interaction per user per hour) and low latency (e.g. 1ms) requirements, and an endpoint solution may not satisfy the requirements. Here we demonstrate how to deploy the policy with SageMaker Batch Transform, where the inferences of the policy are computed beforehand and stored offline in a lookup table for serving. Batch Transform processes input to the policy in batches and is ideal for scenarios where the feedback (in the form of next state and reward) from the environment is expected to be delayed by hours.</blockquote><p> </p><p>I don't think Batch Transform can satisfy low latency requirements. Also, Batch Transform does not have a lookup table stored for serving. And I also understand why batch transform is ideal for delayed rewards in hours</p>

Reply:
> > We may want to evaluate the policy learned from historical data before deployment. Since simulators may not be available in all use cases, we need to evaluate how good the learned policy is using held out historical data. This is called as off-policy evaluation or counterfactual evaluation. We evaluate the policy using several off-policy evaluation algorithms in this notebook.
> 
> This might create confusion to the audience. The OPE is included as part of the training job in the notebook, while the evaluation job running in the notebook is actually interacting with the simulator, generating rewards but not OPE metrics

This is a good point. I think OPE is important because customer may not be able to do the sim evaluation. We should highlight the OPE part in the training job of the notebook and show policy improvement as measured by OPE. We can then explain that we do "real evaluation" with the sim, and show that the policy has improved and the improvement corresponds well to the improvements shown by OPE.

Reply:
> > We can deploy the policy using SageMaker Hosting endpoint. However, some use cases may have low throughput (e.g., one interaction per user per hour) and low latency (e.g. 1ms) requirements, and an endpoint solution may not satisfy the requirements. Here we demonstrate how to deploy the policy with SageMaker Batch Transform, where the inferences of the policy are computed beforehand and stored offline in a lookup table for serving. Batch Transform processes input to the policy in batches and is ideal for scenarios where the feedback (in the form of next state and reward) from the environment is expected to be delayed by hours.
> 
> I don't think Batch Transform can satisfy low latency requirements. Also, Batch Transform does not have a lookup table stored for serving. And I also understand why batch transform is ideal for delayed rewards in hours

Batch Transform is used to pre-compute the inferences and store in a lookup database like Elasticache. The low latency is satisfied by the fast local lookup of Elasticache. 

Batch Transform will not be able to handle fast state transitions, as each state transition will be separate job. Since the spin up time of Batch Transform is of the order of minutes, state transitions of order of minutes or less cannot be supported. Slower state transitions (hence "one interaction per user per hour") can be easily supported. If the state transitions are faster than that, it is better to use a hosting endpoint.

Reply:
<blockquote>This is a good point. I think OPE is important because customer may not be able to do the sim evaluation. We should highlight the OPE part in the training job of the notebook and show policy improvement as measured by OPE. We can then explain that we do "real evaluation" with the sim, and show that the policy has improved and the improvement corresponds well to the improvements shown by OPE.</blockquote><p>Yes, we do show OPE plotting in the notebook and shows the improvement during training. We also do 'real evaluation' with sim and compare with the baseline. We just didn't mention this in the beginning of the notebook. For 'real evaluation', we can only show number of accumulative rewards, we are unable to to show the convergence like that in the tutorial. In order to plot that, we need to load 100 checkpoints and do 100 evaluation jobs, which is not feasible in the notebook</p>

Reply:
<blockquote>Batch Transform is used to pre-compute the inferences and store in a lookup database like Elasticache. The low latency is satisfied by the fast local lookup of Elasticache.</blockquote><blockquote>Batch Transform will not be able to handle fast state transitions, as each state transition will be separate job. Since the spin up time of Batch Transform is of the order of minutes, state transitions of order of minutes or less cannot be supported. Slower state transitions (hence "one interaction per user per hour") can be easily supported. If the state transitions are faster than that, it is better to use a hosting endpoint.</blockquote><p> </p><p>This is definitely one way to use predictions from Batch Transform, but it's not offer of Batch Transform. Doing Batch Transform does not enable storing in lookup table and thus low latency. Elasticcache is the thing to enable this, but we are not mentioning or using it anywhere in the notebook.</p><p> </p><p>Even the state transitions are less than minutes, we can still batch transform. If we are interacting sims in the notebook, all the state transitions will be finished in seconds, and that doesn't blocking batch transform.</p><p> </p><p> </p>